raft/raft.go

```diff
@@ -1171,33 +1213,58 @@ func (r *raft) restore(s pb.Snapshot) bool {
 
 	r.raftLog.restore(s)
 	r.prs = make(map[uint64]*Progress)
-	for _, n := range s.Metadata.ConfState.Nodes {
+	r.restoreNode(s.Metadata.ConfState.Nodes, false)
+	r.restoreNode(s.Metadata.ConfState.Learners, true)
+	return true
+}
+
+func (r *raft) restoreNode(nodes []uint64, isLearner bool) {
+	for _, n := range nodes {
 		match, next := uint64(0), r.raftLog.lastIndex()+1
 		if n == r.id {
 			match = next - 1
+			r.isLearner = isLearner
 		}
-		r.setProgress(n, match, next)
+		r.setProgress(n, match, next, isLearner)
 		r.logger.Infof("%x restored progress of %x [%s]", r.id, n, r.prs[n])
 	}
-	return true
 }
 
 // promotable indicates whether state machine can be promoted to leader,
-// which is true when its own id is in progress list.
+// which is true when its own id is in progress list and its not learner.
 func (r *raft) promotable() bool {
 	_, ok := r.prs[r.id]
-	return ok
+	return ok && !r.isLearner
 }
 
 func (r *raft) addNode(id uint64) {
+	r.addLearnerNode(id, false)
+}
+
+func (r *raft) addLearner(id uint64) {
+	r.addLearnerNode(id, true)
+}
+
+func (r *raft) addLearnerNode(id uint64, isLearner bool) {
```

Comment:
addNodeOrLearnerNode. 

or probably just duplicate code a little bit, and remove this func.